### ISSUE TITLE
eval_objective, eval_grad, step and overall class structure in pyro.infer

### ISSUE THREAD

**Issue by dustinvtran** [pyro-ppl/pyro#168]
I'm not sure how the inference API will be structured overall, but here are some comments for existing VI algorithms:

+ `KL_QP.eval_grad` returns a surrogate loss whose autodiff are gradients for the optimizer. `KL_QP.eval_objective` returns the loss function. `KL_QP.step` calls `eval_grad`, runs autodiff and the update, and returns the loss from `KL_QP.eval_grad`.
+ If `eval_grad` is made to be exposed to the user, returning the surrogate loss as opposed to explicit gradients can be confusing, and also more limiting for general VI algorithms. However, maybe there's a performance concern in that Pytorch is slower when explicitly handling the gradients?
+ `step` should consistently return the objective to minimize or as the opposite standard, the objective to maximize. `KL_QP.step` currently returns the -ELBO and `TraceGraph_KL_QP` returns the ELBO. I suggest returning the objective to minimize. (This also justifies `KL_QP` as the name over `ELBO`, since we're minimizing said objective :))
+  In an example such as score function gradients, `KL_QP.step` would have to call `KL_QP.eval_objective` to return the loss. This seems expensive since computation is re-used across `eval_objective` and `eval_grad`; so maybe they should be combined?
